base/dates/io-util.jl

```diff
@@ -0,0 +1,178 @@
+### Parsing utilities
+
+macro chk1(expr,label=:error)
+    quote
+        x = $(esc(expr))
+        if isnull(x[1])
+            @goto $label
+        else
+            get(x[1]),x[2]
+        end
+    end
+end
+
+
+@generated function _tryparse{T, N}(fmt::DateFormat{T, NTuple{N}}, str::AbstractString)
+    quote
+        R = Nullable{NTuple{7,Int}}
+        t = fmt.tokens
+        l = fmt.locale
+        len = endof(str)
+
+        state = start(str)
+        err_idx = 1
+        Base.@nexprs $N i->val_i = 0
+        Base.@nexprs $N i->(begin
+            state > len && @goto done
+            (val_i, state) = @chk1 tryparsenext(t[i], str, state, len, l)
+            err_idx += 1
+        end)
+        state <= len && @goto error
+
+        @label done
+        parts = Base.@ntuple $N val
+        return R(reorder_args(parts, fmt.field_order, fmt.field_defaults, err_idx)::NTuple{7,Int})
+
+        @label error
+        return R((err_idx,state,0,0,0,0,0), true)
+    end
+end
```

Comment:
I know the classic way of doing it in a static language (hardcoding a jump table / switch / vtable), but afaik there isn't really a good way to express that optimization in Julia right now (other than by doing a generated function like this). fwiw, since this only uses `@nexpr`, it appears to be a completely valid generated function.